### PR TITLE
 Use reactive streams sync adapters to share AbstractClientSideEncryptionTest

### PR DESCRIPTION
**Pull request by jyemin**
https://jira.mongodb.org/browse/JAVA-3393
https://jira.mongodb.org/browse/JAVA-3394

### CODE REVIEW

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoClient.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        return new SyncClientSession(requireNonNull(subscriber.first()), this);
+    }
+
+    private <TResult> SyncSubscriber<TResult> createSubscriber() {
+        return new SyncSubscriber<>();
+    }
+
+    @Override
+    public ClientSession startSession(final ClientSessionOptions options) {
+        throw new UnsupportedOperationException();
```
Comment (jyemin) [inline driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoClient.java]: One example of where the adapter is not fully implemented.  Will fill these in as required by subsequent test sharing.

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCursor.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import com.mongodb.client.MongoCursor;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class SyncMongoCursor<T> implements MongoCursor<T> {
```
Comment (jyemin) [inline driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCursor.java]: This was the trickiest bit.

Comment (rozza) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCursor.java]: crafty

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCollection.java [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+    }
+
+    @Override
+    public <TResult> FindIterable<TResult> find(final ClientSession clientSession, final Bson filter, final Class<TResult> resultClass) {
+        return new SyncFindIterable<>(wrapped.find(unwrap(clientSession), filter, resultClass));
+    }
+
+    @Override
+    public AggregateIterable<T> aggregate(final List<? extends Bson> pipeline) {
+        // TODO: mismatch between sync and async APIs
```
Comment (rozza) [inline driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCollection.java]: How are these mismatched?

Comment (jyemin) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCollection.java]: In the sync API the return type is `AggregateIterable<Document>`

**Pull request closed without merging**